kernel/constr.ml

```diff
@@ -491,6 +491,18 @@ let map_with_binders g f l c0 = match kind c0 with
     let bl' = CArray.Fun1.smartmap f l' bl in
     mkCoFix (ln,(lna,tl',bl'))
 
+type instance_compare_fn = global_reference -> int ->
+  Univ.Instance.t -> Univ.Instance.t -> bool
+
+(* [upto [kind], break [c] into a head [kind] and arguments ignoring casts. *)
+let rec break_app kind c =
```

Comment:
That looks like collapse_appl in termops.ml

Reply:
`collapse_appl sigma c = EConstr.mkApp (Constr.break_app (Econstr.kind sigma) (EConstr.unsafe_of_constr c)` except that it does some shortcut preserving sharing.

Also I don't think EConstr has the property that if `kind c = App (hd, args)` then `hd` itself is not an `App`, so that could lead to subtle evar sensitivity bugs.

Reply:
@SkySkimmer Yes it does, at least at a given unification state. The only weird thing that can happen is that if `hd` is an evar, and *then* some unification occurs, then it may be the case that `hd` is expanded into an application. But for a fixed sigma this invariant holds, and similar ones for casts.

Reply:
Indeed looking closer it does whd_evar going under apps and catss.